Refactored, tested, and Pythonised cpdFmax2pCp2.m

j-emberton opened the pull request:
Refactored cpdFmax2pCp2.m using LLM.

- [x] Refactored and tested cpdFmax2pCp2
- [x] Abstracted common for cpdFmax2pCp2 and cpdFmax2pCp3 into cpdFmax2pCore
- [x] Tests and Cp3 updated to common core
- [x] cpdFmax2pCp2 pythonised (using the common core)

j-emberton committed:
Refactored cpdFmax2pCp2.m

j-emberton committed:
Added unit tests in test_cpdFmaxp2Cp2.py

j-emberton committed:
Define unit tests for sub functions in test_cpdFmaxp2Cp2_subfunctions.py

j-emberton committed:
added matlab engine script in conftest.py

casterkay commented on oneflux_steps/ustar_cp/cpdFmax2pCp2.m:
```diff
@@ -1,52 +1,79 @@
+    if Fmax < FmaxCritical(1)
+        p = extrapolateLowFmax(Fmax, n, FmaxCritical(1));
+    elseif Fmax > FmaxCritical(end)
+        p = extrapolateHighFmax(Fmax, n, FmaxCritical(end));
```
In the original code, `FmaxCritical(3)` is used instead of `FmaxCritical(end)`. Are they the same?

dorchard commented:
We will come back to this once we reach this point in the dependency graph.

dorchard commented:
I'm on this now to see if I can sync it up with the work so far and perhaps even reused `cpdFmax2cpCp3`.

dorchard committed:
Merge branch 'ustar_cp_refactor_main' into refactor_and_test_cpdFmax2pCp2

dorchard committed:
convert tests to new test_engine runner

dorchard committed:
abstract Fmax2pCp2 and Cp3 into a single core

dorchard committed:
Cp2 module in python using the common core

dorchard commented:
All tests working correctly. Updated the PR description to reflect that this pythonises Cp2 via an abstracted, parameterised common core for Cp3 and Cp2 change point analysis k-parameter model.

dorchard committed:
nits

dorchard committed:
Merge branch 'ustar_cp_refactor_main' into refactor_and_test_cpdFmax2pCp2

dorchard committed:
Merge branch 'ustar_cp_refactor_main' into refactor_and_test_cpdFmax2pCp2

j-emberton committed:
added new args for cpdFmax* core functions

j-emberton committed:
Merge branch 'refactor_and_test_cpdFmax2pCp2' of github.com:Cambridge-ICCS/ONEFlux into refactor_and_test_cpdFmax2pCp2

j-emberton committed:
removed errant import

j-emberton committed:
removed unnecessary changes to ustar_cp/cpdFmax2pCp2

j-emberton commented:
I've made some fixes to this PR:
- propagated Doms changes into the matlab code
- removed errant import statement that was preventing tests running
- removed changes to ustar_cp/cpdFmax2pCp2 code

All tests pass.

I think this is ready to go

dorchard commented:
Super- thanks!
